File: mobile-app/android/app/src/main/java/com/ensadvantage/app/DeviceIntegrityPlugin.java
package com.ensadvantage.app;

import android.app.ActivityManager;
import android.content.Context;
import android.content.pm.PackageManager;
import android.os.Build;
import android.os.Debug;

import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

import java.io.File;

/**
 * ═══════════════════════════════════════════════════════════════
 * DEVICE INTEGRITY (Android) — root / tamper detection
 * ───────────────────────────────────────────────────────────────
 * Registered as a LOCAL Capacitor plugin — no separate Gradle
 * module, no npm package, nothing to sync — just this one class,
 * annotated with @CapacitorPlugin and registered by name from
 * MainActivity.onCreate() (see the registerPlugin() call added
 * there). This is the standard, documented way to add a small
 * custom native plugin to a Capacitor app without ejecting or
 * publishing a package.
 *
 * Exposes isCompromised() to JS as
 * `window.Capacitor.Plugins.DeviceIntegrity.isCompromised()`,
 * consumed by js/deviceIntegrity.js — that file already falls back
 * to a much weaker JS-only heuristic if this plugin call ever fails
 * or isn't present, so nothing breaks if this file isn't wired up
 * yet on a given build.
 *
 * WHAT THIS ACTUALLY CHECKS: standard, widely-used, well-documented
 * heuristics — no single one of them is proof of root by itself
 * (custom ROMs and some legitimate OEM builds can trip one or two),
 * which is why isCompromised() only returns true once at least 2
 * independent signals agree, exactly mirroring the conservative
 * threshold already used by the JS-only fallback tier.
 *
 * HONEST LIMIT, stated once here: these are the same categories of
 * check every mainstream root-detection library (RootBeer, SafetyNet
 * predecessors, etc.) uses, and every one of them is defeatable by a
 * sufficiently motivated attacker — Magisk's DenyList/Zygisk hiding,
 * a repackaged build with this very check patched out, or a modified
 * ROM that never trips any of these paths at all. This raises the
 * cost of casual tampering meaningfully (it is real OS-level
 * evidence, not a spoofable JS variable); it is not, and cannot be
 * marketed as, an unbreakable barrier. Google's own attestation API
 * (Play Integrity) is the actual state of the art for this and would
 * be a stronger foundation than this file if hardware-backed
 * attestation is ever worth the added Play Services dependency and
 * server-side verification — deliberately out of scope here to keep
 * this addable without any new server component or SDK.
 *
 * ⚠️ BUILD/TEST CAVEAT: this file was written to the correct,
 * standard Capacitor 7 local-plugin shape, but has not been compiled
 * or run — this environment has no Android SDK/emulator/device to
 * verify against. Before shipping any build that GATES functionality
 * on isCompromised() (see materialCache.js), test on at least one
 * real non-rooted device to confirm it reports `rooted: false` and
 * doesn't false-positive — a false positive here silently costs a
 * paying student their offline access, which is a much worse failure
 * mode than under-detecting.
 * ═══════════════════════════════════════════════════════════════
 */
@CapacitorPlugin(name = "DeviceIntegrity")
public class DeviceIntegrityPlugin extends Plugin {

  // Common su binary locations across mainstream root solutions
  // (Magisk, SuperSU, older su-based root methods).
  private static final String[] SU_PATHS = {
    "/system/bin/su",
    "/system/xbin/su",
    "/sbin/su",
    "/system/su",
    "/su/bin/su",
    "/system/bin/.ext/.su",
    "/system/usr/we-need-root/su-backup",
    "/data/local/xbin/su",
    "/data/local/bin/su",
    "/data/local/su",
    "/cache/su",
    "/vendor/bin/su"
  };

  // Packages installed by mainstream root-management / root-hiding tools.
  private static final String[] ROOT_PACKAGES = {
    "com.topjohnwu.magisk",
    "eu.chainfire.supersu",
    "com.noshufou.android.su",
    "com.koushikdutta.superuser",
    "com.thirdparty.superuser",
    "com.yellowes.su",
    "com.zachspong.temprootremovejb",
    "com.ramdroid.appquarantine"
  };

  @PluginMethod
  public void isCompromised(PluginCall call) {
    JSArray signals = new JSArray();

    if (checkSuBinaries())         signals.put("su_binary_present");
    if (checkRootPackages())       signals.put("root_management_app_installed");
    if (checkTestKeysBuildTag())   signals.put("test_keys_build_tag");
    if (checkDebuggerAttached())   signals.put("debugger_attached");
    if (checkEmulator())           signals.put("emulator_detected");

    // Conservative threshold, matching js/deviceIntegrity.js's JS-only
    // tier: a single weak/ambiguous signal (e.g. test-keys on some
    // legitimate OEM/custom-ROM builds, or an emulator used by a
    // developer legitimately testing the app) never blocks a real
    // student by itself. Two or more independent signals agreeing is
    // treated as an actual finding.
    boolean rooted = signals.length() >= 2;

    JSObject result = new JSObject();
    result.put("rooted", rooted);
    result.put("signals", signals);
    call.resolve(result);
  }

  private boolean checkSuBinaries() {
    for (String path : SU_PATHS) {
      if (new File(path).exists()) return true;
    }
    return false;
  }

  private boolean checkRootPackages() {
    PackageManager pm = getContext().getPackageManager();
    for (String pkg : ROOT_PACKAGES) {
      try {
        pm.getPackageInfo(pkg, 0);
        return true; // no exception means the package is installed
      } catch (PackageManager.NameNotFoundException ignored) {
        // expected for every package NOT installed — keep checking the rest
      }
    }
    return false;
  }

  private boolean checkTestKeysBuildTag() {
    // Official, retail Android builds are signed with "release-keys".
    // "test-keys" means the build was signed with the AOSP test
    // certificate — the standard signal for a custom/dev build,
    // though a small number of legitimate custom ROMs also carry
    // this tag, which is exactly why this alone never gates anything
    // (see the >= 2 threshold above).
    String tags = Build.TAGS;
    return tags != null && tags.contains("test-keys");
  }

  private boolean checkDebuggerAttached() {
    return Debug.isDebuggerConnected() || Debug.waitingForDebugger();
  }

  private boolean checkEmulator() {
    // Standard emulator fingerprint checks — an emulator isn't
    // "rooted" in the traditional sense, but it's a device class this
    // app's offline-content protections were never meant to run
    // meaningfully on (an emulator disk image is trivially inspectable
    // regardless of what happens inside the app), so it's folded into
    // the same signal set at the same conservative threshold.
    return (Build.FINGERPRINT != null && (Build.FINGERPRINT.startsWith("generic")
              || Build.FINGERPRINT.startsWith("unknown")
              || Build.FINGERPRINT.contains("test-keys")))
        || (Build.MODEL != null && (Build.MODEL.contains("google_sdk")
              || Build.MODEL.contains("Emulator")
              || Build.MODEL.contains("Android SDK built for x86")))
        || (Build.MANUFACTURER != null && Build.MANUFACTURER.contains("Genymotion"))
        || (Build.BRAND != null && Build.BRAND.startsWith("generic") && Build.DEVICE != null && Build.DEVICE.startsWith("generic"))
        || "google_sdk".equals(Build.PRODUCT);
  }
}

File: mobile-app/ios/App/App/DeviceIntegrityPlugin.swift
import Foundation
import UIKit
import Capacitor

/**
 * ═══════════════════════════════════════════════════════════════
 * DEVICE INTEGRITY (iOS) — jailbreak / tamper detection
 * ───────────────────────────────────────────────────────────────
 * Local Capacitor plugin using the modern (Capacitor 3+) pure-Swift
 * `CAPBridgedPlugin` protocol — no separate Objective-C bridge (.m)
 * file needed, unlike older Capacitor plugin templates. Registered
 * from SecureViewController.capacitorDidLoad() (see that file) since
 * a locally-bundled plugin like this one isn't auto-discovered the
 * way an npm-installed plugin's podspec is.
 *
 * Exposes isCompromised() to JS as
 * `window.Capacitor.Plugins.DeviceIntegrity.isCompromised()`,
 * consumed by js/deviceIntegrity.js — that file already falls back
 * to a much weaker JS-only heuristic if this plugin call ever fails
 * or isn't present, so nothing breaks if this file isn't wired up
 * yet on a given build.
 *
 * WHAT THIS CHECKS — the same standard, widely-documented jailbreak
 * heuristics used by mainstream detection libraries (IOSSecuritySuite,
 * DTTJailbreakDetection, etc.):
 *   - Well-known jailbreak file paths (Cydia, Substrate, common
 *     jailbreak-tool binaries)
 *   - Ability to write outside the app's sandbox (only possible with
 *     jailbreak-granted elevated filesystem permissions)
 *   - The `cydia://` URL scheme being openable (requires the
 *     `LSApplicationQueriesSchemes` entry added to Info.plist
 *     alongside this file)
 *   - A debugger currently attached, via the classic sysctl
 *     P_TRACED flag check
 *
 * Same conservative threshold as the Android plugin and the JS
 * fallback tier: isCompromised() only returns true once at least 2
 * independent signals agree, so no single ambiguous signal (some
 * enterprise MDM configurations, some legitimate low-level dev/debug
 * sessions) locks out a real student by itself.
 *
 * HONEST LIMIT, stated once here: every one of these checks is
 * defeatable by modern jailbreak-hiding tools (Shadow, A-Bypass,
 * etc.) or a repackaged IPA with this very check patched out. This
 * raises the cost of casual tampering — it is real OS-level
 * evidence, not a spoofable JS variable — but it is not, and cannot
 * be marketed as, an unbreakable barrier. Apple's DeviceCheck / App
 * Attest APIs are the actual state of the art here and would be a
 * stronger foundation if server-side hardware attestation is ever
 * worth the added complexity — deliberately out of scope here to
 * keep this addable without any new server component.
 *
 * ⚠️ BUILD/TEST CAVEAT: written to the correct, standard Capacitor 7
 * CAPBridgedPlugin shape, but not compiled or run — this environment
 * has no Xcode/simulator/device to verify against, and (see the repo
 * note handed back alongside this file) this .swift file still needs
 * to be added to the Xcode project itself, since raw file drops don't
 * register with Xcode's project.pbxproj automatically. Test on at
 * least one real non-jailbroken device before shipping any build that
 * GATES functionality on isCompromised() (see materialCache.js) — a
 * false positive here silently costs a paying student their offline
 * access, which is a much worse failure mode than under-detecting.
 * ═══════════════════════════════════════════════════════════════
 */
@objc(DeviceIntegrityPlugin)
public class DeviceIntegrityPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "DeviceIntegrityPlugin"
    public let jsName = "DeviceIntegrity"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "isCompromised", returnType: CAPPluginReturnPromise)
    ]

    private static let jailbreakPaths = [
        "/Applications/Cydia.app",
        "/Applications/Sileo.app",
        "/Applications/Zebra.app",
        "/Library/MobileSubstrate/MobileSubstrate.dylib",
        "/Library/MobileSubstrate/DynamicLibraries",
        "/bin/bash",
        "/usr/sbin/sshd",
        "/etc/apt",
        "/private/var/lib/apt",
        "/private/var/lib/cydia",
        "/private/var/stash",
        "/var/checkra1n.dmg",
        "/usr/libexec/cydia/firmware.sh"
    ]

    @objc func isCompromised(_ call: CAPPluginCall) {
        var signals: [String] = []

        if Self.checkJailbreakFiles()   { signals.append("jailbreak_file_present") }
        if Self.checkSandboxEscape()    { signals.append("sandbox_write_succeeded") }
        if Self.checkCydiaUrlScheme()   { signals.append("cydia_url_scheme_openable") }
        if Self.checkDebuggerAttached() { signals.append("debugger_attached") }
        if Self.checkSimulator()        { signals.append("simulator_detected") }

        // Conservative threshold — see file header.
        let compromised = signals.count >= 2

        call.resolve([
            "rooted": compromised,
            "signals": signals
        ])
    }

    private static func checkJailbreakFiles() -> Bool {
        let fm = FileManager.default
        return jailbreakPaths.contains { fm.fileExists(atPath: $0) }
    }

    private static func checkSandboxEscape() -> Bool {
        // A non-jailbroken app's sandbox forbids writing outside its
        // own container — only jailbreak-granted elevated filesystem
        // access lets this succeed.
        let testPath = "/private/ens_integrity_check_\(UUID().uuidString).txt"
        do {
            try "test".write(toFile: testPath, atomically: true, encoding: .utf8)
            try? FileManager.default.removeItem(atPath: testPath) // clean up regardless of what caused the write to succeed
            return true
        } catch {
            return false // expected outcome on a normal, non-jailbroken device
        }
    }

    private static func checkCydiaUrlScheme() -> Bool {
        // Requires "cydia" listed under LSApplicationQueriesSchemes in
        // Info.plist (added alongside this file) — without it,
        // canOpenURL always returns false regardless of whether Cydia
        // is actually installed, which would make this check useless
        // rather than merely conservative.
        guard let url = URL(string: "cydia://package/com.example.package") else { return false }
        var result = false
        if Thread.isMainThread {
            result = UIApplicationBridge.canOpen(url)
        } else {
            DispatchQueue.main.sync { result = UIApplicationBridge.canOpen(url) }
        }
        return result
    }

    private static func checkDebuggerAttached() -> Bool {
        // Classic sysctl-based debugger check: ask the kernel for this
        // process's own info and inspect the P_TRACED flag, the same
        // technique Apple's own sample code and every mainstream
        // anti-debug library on iOS uses.
        var info = kinfo_proc()
        var mib: [Int32] = [CTL_KERN, KERN_PROC, KERN_PROC_PID, getpid()]
        var size = MemoryLayout<kinfo_proc>.stride
        let result = sysctl(&mib, UInt32(mib.count), &info, &size, nil, 0)
        guard result == 0 else { return false }
        return (info.kp_proc.p_flag & P_TRACED) != 0
    }

    private static func checkSimulator() -> Bool {
        #if targetEnvironment(simulator)
        return true
        #else
        return false
        #endif
    }
}

/// Tiny indirection so canOpenURL (a UIApplication instance method,
/// which must be called on the main thread) has a single, clearly-
/// named call site above rather than reaching into UIApplication.shared
/// inline in two branches.
private enum UIApplicationBridge {
    static func canOpen(_ url: URL) -> Bool {
        UIApplication.shared.canOpenURL(url)
    }
}
